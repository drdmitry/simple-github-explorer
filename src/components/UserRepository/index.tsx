import type { FC } from "react";

import type { IUserRepository } from "~/interfaces/user-repository.interface";

import SvgGitHub from "~/components/Icons/SvgGitHubIcon";
import SvgStar from "~/components/Icons/SvgStarIcon";
import SvgIcon, { ICON_NAMES } from "../Icons/SvgIcon";

interface IProps {
  repo: IUserRepository;
};

export const UserRepository: FC<IProps> = ({ repo }) => {
  const handleRepoClick = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    event.preventDefault();
    window.open(repo.html_url, "_blank");
  };

  return (
    <div className="w-full flex flex-col p-3 rounded-md bg-slate-300 min-w-[200px] shadow-sm">
      <div className="flex flex-col min-[370px]:flex-row align-center justify-between">
        <div className="text-lg font-bold text-gray-900">
          {repo.name}
        </div>
        <div className="text-sm text-gray-500 self-center flex flex-row gap-1 items-center">
          {repo.stargazers_count}
          <SvgIcon name={ICON_NAMES.Star} size={16} />
        </div>
      </div>
      <div className="text-sm text-gray-500">
        {repo.description}
      </div>
      <div className="h-2" />
      <div
        className="flex flex-row items-center place-self-end text-sm text-gray-500 gap-2 hover:text-gray-700 cursor-pointer overflow-x-hidden"
        onClick={handleRepoClick}
        title="opens in new browser tab"
      >
        <SvgGitHub size={20} />
        {repo.full_name}
      </div>
    </div>
  );
};

export default UserRepository;
