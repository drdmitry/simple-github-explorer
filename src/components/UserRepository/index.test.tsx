import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import UserRepository from './';

import type { IUserRepository } from '~/interfaces/user-repository.interface';

// Mock the repository data for testing
const mockRepo: IUserRepository = {
  id: 1,
  name: 'Test Repo',
  stargazers_count: 42,
  description: 'A test repository',
  html_url: 'https://github.com/testuser/test-repo',
  full_name: 'testuser/test-repo',
  private: false,
  created_at: '2020-01-01T00:00:00Z',
  updated_at: '2020-01-01T00:00:00Z',
};

// Function to render the UserRepository component
function renderUserRepository() {
  return render(<UserRepository repo={mockRepo} />);
}

describe("UserRepository", () => {
  test('renders repository name correctly', () => {
    const { getByText } = renderUserRepository();
    const repoNameElement = getByText('Test Repo');
    expect(repoNameElement).toBeInTheDocument();
  });

  test('renders star count and star icon correctly', () => {
    const { getByText, container } = renderUserRepository();
    const starCountElement = getByText('42');
    expect(starCountElement).toBeInTheDocument();
    const starIconElement = container.querySelector('svg');
    expect(starIconElement).toBeInTheDocument();
  });

  test('renders repository description correctly', () => {
    const { getByText } = renderUserRepository();
    const descriptionElement = getByText('A test repository');
    expect(descriptionElement).toBeInTheDocument();
  });

  test('opens repository link in a new tab when clicked', () => {
    const spyWindowOpen = jest.spyOn(window, 'open');
    spyWindowOpen.mockImplementation(jest.fn());
    const { getByText } = renderUserRepository();
    const repositoryLinkElement = getByText('testuser/test-repo');
    fireEvent.click(repositoryLinkElement);
    expect(spyWindowOpen).toHaveBeenCalledWith('https://github.com/testuser/test-repo', '_blank');
  });

  test("renders UserRepository unchanged", () => {
    const { container } = renderUserRepository();
    expect(container).toMatchSnapshot();
  });
});
