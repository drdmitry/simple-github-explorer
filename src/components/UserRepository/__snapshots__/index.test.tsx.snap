// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRepository renders UserRepository unchanged 1`] = `
<div>
  <div
    class="w-full flex flex-col p-3 rounded-md bg-slate-300 min-w-[200px] shadow-sm"
  >
    <div
      class="flex flex-col min-[370px]:flex-row align-center justify-between"
    >
      <div
        class="text-lg font-bold text-gray-900"
      >
        Test Repo
      </div>
      <div
        class="text-sm text-gray-500 self-center flex flex-row gap-1 items-center"
      >
        42
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 20 20"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      class="text-sm text-gray-500"
    >
      A test repository
    </div>
    <div
      class="h-2"
    />
    <div
      class="flex flex-row items-center place-self-end text-sm text-gray-500 gap-2 hover:text-gray-700 cursor-pointer overflow-x-hidden"
      title="opens in new browser tab"
    >
      <svg
        height="20"
        viewBox="0 0 98 98"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M48.854 0C21.839 0 0 22 0 49.217c0 21.756 13.993 40.172 33.405 46.69 2.427.49 3.316-1.059 3.316-2.362 0-1.141-.08-5.052-.08-9.127-13.59 2.934-16.42-5.867-16.42-5.867-2.184-5.704-5.42-7.17-5.42-7.17-4.448-3.015.324-3.015.324-3.015 4.934.326 7.523 5.052 7.523 5.052 4.367 7.496 11.404 5.378 14.235 4.074.404-3.178 1.699-5.378 3.074-6.6-10.839-1.141-22.243-5.378-22.243-24.283 0-5.378 1.94-9.778 5.014-13.2-.485-1.222-2.184-6.275.486-13.038 0 0 4.125-1.304 13.426 5.052a46.97 46.97 0 0 1 12.214-1.63c4.125 0 8.33.571 12.213 1.63 9.302-6.356 13.427-5.052 13.427-5.052 2.67 6.763.97 11.816.485 13.038 3.155 3.422 5.015 7.822 5.015 13.2 0 18.905-11.404 23.06-22.324 24.283 1.78 1.548 3.316 4.481 3.316 9.126 0 6.6-.08 11.897-.08 13.526 0 1.304.89 2.853 3.316 2.364 19.412-6.52 33.405-24.935 33.405-46.691C97.707 22 75.788 0 48.854 0z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      testuser/test-repo
    </div>
  </div>
</div>
`;
